static/rfcs/17/rln-relay.msc

```diff
@@ -1,32 +1,23 @@
 # Sequence diagram for RLN Relay protocol (registration)
 msc {
-  hscale = "2";
+  hscale = "1";
 
   d [label = " "], a [label = " "],b [label = " "];
   a rbox a [label="Relay Node A"], 
   b note b [label="Membership Contract"],
   d rbox d [label = "Relay Node B"];
+ 
   |||;
+  d abox d [ label=" \n Listening to the membership contract \n "] ;
 
   a box a [ label=" \n Generate sk,pk \n "] ;
   a=>b [ label = " \n Register(pk, x ETH) \n " ] ;
-  |||;
+  a abox a [ label=" \n Listening to the membership contract \n "] ;
 
-  b box b [label=" \n Insert pk as a leaf to the tree \n \n index: The index of the inserted leaf \n \n root: The updated tree root \n \n authPath: The authentication path \n  "];
-  |||;
-  b=>a [ label = "index, root, authPath"];
-  |||;
-  ...,  ---;
-  ...   [ label = "Other relay nodes register and the membership tree gets updatetd" ];
-  ...,  ---;
-  a=>b [ label = "getRoot()" ] ;
-  b box b [label="  \n  root: Get the current root\n  "];
-  b=>a [ label = "root"];
+  b box b [label=" \n Insert pk  to the list. \n Emit an event announcing the insertion of pk and its index in the list. \n  "];
 
-  ...,  ---  [ label = "    " ];
+  b=>a [ label = "Insert(pk, index)"];
+  b=>d [ label = "Insert(pk, index)"];
```

Comment:
The contract should return (pk, index) to nodes to allow them to update their local trees (I assumed so, since you removed the getAuthPath/getRoot calls).

Reply:
IIRC contract won't return this, since transaction hasn't been mined yet, but you have to listen for it to be emitted as an event.

See e.g. https://github.com/ChihChengLiang/poseidon-tornado/blob/main/contracts/Tornado.sol#L66 and https://docs.ethers.io/v5/concepts/events/

Reply:
>IIRC contract won't return this, since transaction hasn't been mined yet, but you have to listen for it to be emitted as an event.

The figure is a simplified version of actual interaction with the contract, and is clearly not precise. I have revised it to better reflect the mining phase as well.  https://github.com/vacp2p/rfc/pull/497/commits/002a81223dd8d8be94eb20abad826b519cf043ed